devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -0,0 +1,663 @@
+
+using namespace hde::devices;
+
+class HumanDataCollector::impl
+{
+public:
+
+    double period;
+    std::string portPrefix;
+    bool firstData = true;
+
+    DataBuffersConversionHelper dataBuffersConversionHelper;
+
+    hde::interfaces::IHumanState* iHumanState = nullptr;
+    hde::interfaces::IHumanWrench* iHumanWrenchMeasurements = nullptr;
+    hde::interfaces::IHumanWrench* iHumanWrenchEstimates = nullptr; //This interface points to HumanDynamicsEstimator, which gives offset removed wrench measurements and the estimates
+    hde::interfaces::IHumanDynamics* iHumanDynamics = nullptr;
+
+    bool isStateProviderDevice = false;
+    bool isWrenchProviderDevice = false;
+    bool isDynamicsEstimatorDevice = false;
+
+    // Interface data buffers
+
+    // Base Quantities
+    std::string baseName;
+    std::array<double, 3> basePosition;
+    std::array<double, 4> baseOrientation;
+    std::array<double, 6> baseVelocity;
+
+    // Joint Quantities
+    size_t stateNumberOfJoints;
+    std::vector<std::string> stateJointNames;
+    std::vector<double> jointPositions;
+    std::vector<double> jointVelocities;
+
+    // CoM Quantities
+    std::array<double, 3> comPosition;
+    std::array<double, 3> comVelocity;
+    std::array<double, 6> comProperAccInBaseFrame;
+    std::array<double, 6> comProperAccInWorldFrame;
+
+    // Wrench Measurements
+    size_t numberOfWrenchMeasurementSources;
+    std::vector<std::string> wrenchMeasurementSourceNames;
+    std::vector<double> wrenchMeasurementValues;
+
+    // Wrench Estimates
+    size_t numberOfWrenchEstimateSources;
+    std::vector<std::string> wrenchEstimateSourceNames;
+    std::vector<double> wrenchEstimateValues;
+
+    // Joint Torques
+    size_t dynamicsNumberOfJoints;
+    std::vector<std::string> dynamicsJointNames;
+    std::vector<double> jointTorques;
+
+    // Yarp ports for streaming data from IHumanState interface of HumanStateProvider
+    yarp::os::BufferedPort<yarp::sig::Vector> basePoseDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> baseVelocityDataPort;
+    yarp::os::BufferedPort<yarp::os::Bottle> stateJointNamesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> jointPositionsDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> jointVelocitiesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> comPositionDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> comVelocityDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> comProperAccelerationInBaseFrameDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> comProperAccelerationInWorldFrameDataPort;
+
+    // Yarp ports for streaming data from IHumanWrench interface of HumanWrenchProvider
+    yarp::os::BufferedPort<yarp::os::Bottle> wrenchMeasurementSourceNamesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> wrenchMeasurementValuesDataPort;
+
+    // Yarp ports for streaming data from IHumanWrench interfce and IHumanDynamics interface of HumanDynamicsEstimator
```

Comment:
Probably we can stream offset-removed-wrench and estimated-wrench on two separate ports, since the first one is not really a relevant quantity while the second one it is.

In this way we would have three ports streaming wrench measurements with the same size (instead of having `wrenchMeasurementValuesDataPort ` with size **n** and `wrenchEstimateValuesDataPort` with size **2n**). The advantage is that then code can be recycled (e.g. reading and plotting the three ports would be structured in the same way for the three ports)

Reply:
... and bottles maybe could be removed from this class :)

Reply:
> ... and bottles maybe could be removed from this class :)

I settled on bottles to stream the names of the joints or the wrench sources. Is there  any better alternative to stream a vector of strings through yarp ports?

Reply:
I don't remember if `yarp::sig::Vector` works with strings :) It's worth checking.

Reply:
@diegoferigo I already went down that road :sweat_smile: and then lazily settled with `Bottle`. May be @traversaro can point to something in this regard.

Reply:
I know that you use Bottles a lot :) In general, if you can use something that can be serialized to a string but provides higher-levels APIs like `yarp::sig::Vector<std::string>` (or, `unsigned char`) you'd get more readable code and less error prone.

Reply:
> eady went down that road sweat_smile and then lazily settled with `Bottle`. May be @traversaro can point to something in this regard.

I do not know if `yarp::sig::VectorOf` works with `std::string`.

Reply:
Compiled without any errors but when I try to read the port, it crashes. Unfortunately I didn’t document the crash log at the moment.